apps/web/src/lib/tiptap/ipfs-image.tsx

```diff
@@ -0,0 +1,182 @@
+import { Node } from "@tiptap/core";
+import { mergeAttributes, NodeViewWrapper, ReactNodeViewRenderer } from "@tiptap/react";
+import { useEffect, useState } from "react";
+import { useQuery, useMutation } from "react-query";
+import { Loader2 } from "lucide-react";
+import { CommandExtensionProps } from "./types";
+import { DataPill } from "../../components/DataPill";
+
+export type IpfsImageProps = {
+  cid: string | null;
+  alt: string;
+  title: string;
+  file: File | undefined;
+};
+
+async function uploadImageToIpfs(file: File): Promise<string> {
+  const formData = new FormData();
+  formData.append("image", file);
+
+  const res = await fetch("/api/editor/upload", {
+    method: "POST",
+    body: formData,
+    headers: {
+      "Accept": "application/json",
+    },
+  });
+
+  if (!res.ok) {
+    throw new Error("Failed to upload image to IPFS");
+  }
+
+  const { uploads } = await res.json();
+
+  const cid = uploads[0].IpfsHash;
+
+  return cid;
+}
+
+export const IpfsImage = (
+  props: CommandExtensionProps<IpfsImageProps>
+) => {
+  const { cid, file, alt, title } = props.node.attrs;
+  const [objectUrl, setObjectUrl] = useState<string | null>(null);
+
+  const fetchImage = useQuery({
+    queryKey: ["ipfs-image", cid],
+    queryFn: async () => {
+      const res = await fetch(`https://gateway.pinata.cloud/ipfs/${cid}`);
+      const imageBlob = await res.blob();
+      return imageBlob;
+    },
+    refetchOnWindowFocus: false,
+    staleTime: 3 * 60 * 1000,
+    cacheTime: 5 * 60 * 1000,
+    enabled: cid !== null,
+  });
+
+  const uploadImage = useMutation({
+    mutationFn: uploadImageToIpfs,
+    onSuccess: (cid) => {
+      props.updateAttributes({
+        cid,
+        file: undefined,
+        alt: file?.name,
+        title: `${file?.name} uploaded to IPFS with CID ${cid}`,
+      });
+    },
+  });
+
+  const isError = uploadImage.isError || fetchImage.isError;
+  const isLoading = uploadImage.isLoading || fetchImage.isLoading || objectUrl === null;
+
+  useEffect(() => {
+    if (file && cid === null && !uploadImage.isLoading) {
+      setObjectUrl(URL.createObjectURL(file));
+      uploadImage.mutate(file);
+    }
+  }, [cid, file, uploadImage]);
+
+  useEffect(() => {
+    if (fetchImage.data) {
+      setObjectUrl(URL.createObjectURL(fetchImage.data));
+    }
+  }, [fetchImage.data]);
+
+  useEffect(() => {
+    return () => {
+      if (objectUrl) {
+        URL.revokeObjectURL(objectUrl);
+      }
+    };
+  }, [objectUrl]);
+
+  if (isError) {
+    return (
+      <NodeViewWrapper as="span">
+        <DataPill query={fetchImage}>no data</DataPill>;
+      </NodeViewWrapper>
+    );
+  }
+
+  return (
+    <NodeViewWrapper as="span">
```

Comment:
Maybe we want it `as="div"`?

Reply:
Not for now, unless we want inline images.